### PR TITLE
Caching with ElastiCache for Redis and Spring cloud AWS

### PR DESCRIPTION
**Pull request by murtuza-ranapur**
Please re-read your article a day or so after you finished writing it and work through this checklist to polish the quality. This will save all of us a lot of time during review.

## Content
- [x] the article explains *why* and not only *how*
- [X] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text

## Text
- [X] the text is conversational ("we" instead of "you", easy language)
- [X] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [X] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [X] the text uses active voice instead of passive voice
- [X] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [X] file names and variables are highlighted as `code`

## Structure
- [X] the article has a short intro before the first headline
- [X] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [X] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [X] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [X] code examples are introduced with a ":" and explained below the code example
- [X] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)
- [X] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

**Feel free to raise the PR even if the checklist hasn't been worked through, yet**, to take advantage of the preview that is generated for each pull request (see "deploy/netlify" under "Show all checks"). But please only flag it to be reviewed after going through the checklist above.

### CODE REVIEW

File: _posts/2021-05-30-spring-elasticache-redis.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    Properties:
+      ClusterName: user-cache
+      Engine: "Redis"
+      EngineVersion: "6.x"
+      CacheNodeType: !Ref InstanceType
+      NumCacheNodes: 1
+      VpcSecurityGroupIds: [ "sg-xxxx" ]
+```
+
+Then instead of giving cluster names, we only need to provide the stack name. Say the stack name
```
Comment (pratikdas) [inline _posts/2021-05-30-spring-elasticache-redis.md:171]: This part is enough to describe this feature.

File: _posts/2021-05-30-spring-elasticache-redis.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+```groovy
+implementation 'com.amazonaws:aws-java-sdk-cloudformation'
+```
+
+
+## How Does Spring Cloud AWS Configure the `CacheManager`?
+
+In this section, we will dive a bit deeper into the inner workings of the Spring Cloud AWS
+and see how it autoconfigures the cache for us.
```
Comment (pratikdas) [inline _posts/2021-05-30-spring-elasticache-redis.md:207]: A diagram will help here.

File: _posts/2021-05-30-spring-elasticache-redis.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+* Also `ElastiCacheFactoryBean` passes this host and port to `RedisCacheFactory` which then uses Redis clients such as Lettuce to create the connection object which then actually establishes with the nodes and perform required operations.
+
+## Conclusion
+
+While ElastiCache is already making our life easier by managing our Redis Clusters, Spring Cloud AWS further simplifies
+our lives by simplifying the configurations required for communicating with it.
+
+In this article, we saw those configurations and also how to apply them. Hope this was helpful!
+
+Thank you for reading! You can find the working code at [GitHub](https://github.com/thombergs/code-examples/tree/master/spring-boot/caching-with-elasticache-redis).
```
Comment (pratikdas) [inline _posts/2021-05-30-spring-elasticache-redis.md:230]: put the code under aws folder and change the url

File: _posts/2021-05-30-spring-elasticache-redis.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+![Spring Cloud AWS](/assets/img/posts/spring-elasticache-redis/spring-cloud-classes.png)
+
+* When our application starts in the AWS environment, `ElastiCacheAutoConfiguration` reads cluster names from the
+  `application.yml` and passes them on to the `ElastiCacheCacheConfigurer` object.
+* Then `ElastiCacheCacheConfigurer` creates the `CacheManager` with the help of `ElastiCacheFactoryBean` class.
+* `ElastiCacheFactoryBean` scans the ElastiCache in the same availability zone and retrieves the host and port
+  names of the nodes. To allow our service to scan ElatiCache we need to provide `AmazonElastiCacheReadOnlyAccess` permission
+  to our service or `AWSCloudFormationReadOnlyAccess` if we are using the stack name approach.
+* Also `ElastiCacheFactoryBean` passes this host and port to `RedisCacheFactory` which then uses Redis clients such as Lettuce to create the connection object which then actually establishes with the nodes and perform required operations.
+
```
Comment (pratikdas) [inline _posts/2021-05-30-spring-elasticache-redis.md:222]: Please relook at the article in totality. At the end of the article the reader should be able to run the application with the cache configured. 
Somewhere I expect to see caching configuration using Java configuration along with creation of a Spring boot project.

**Review by pratikdas — CHANGES_REQUESTED** [review]

File: _posts/2021-05-30-spring-elasticache-redis.md (hunk 2 above)
Comment (murtuza-ranapur) [reply _posts/2021-05-30-spring-elasticache-redis.md:207]: This section already has a class diagram, is it not informative? Do you think a flow chart would be better?

File: _posts/2021-05-30-spring-elasticache-redis.md (hunk 4 above)
Comment (murtuza-ranapur) [reply _posts/2021-05-30-spring-elasticache-redis.md:222]: Sure, will do that. I have omitted one configuration which is important for running the application will add that too.

For configuring cache apart from auto configuration the only other approach available with Spring Cloud is the annotations based but they have deprecated that. If I have to showcase Java configuration that would mean going back to the non spring cloud approach where we configure caching factories and cache managers by ourselves. Please let me know if I didn't get your point. 

Also, can you please elaborate the `creation of spring boot project` point?

File: _posts/2021-05-30-spring-elasticache-redis.md (hunk 2 above)
Comment (pratikdas) [reply _posts/2021-05-30-spring-elasticache-redis.md:207]: yes. The sequence of interactions is not clear in the class diagram. The text can also be broken down in bullets.

File: _posts/2021-05-30-spring-elasticache-redis.md (hunk 4 above)
Comment (pratikdas) [reply _posts/2021-05-30-spring-elasticache-redis.md:222]: Where will the reader start from? You need to mention that step.

"The caching support for Spring Cloud AWS  uses Spring Data Redis for Redis caches". So you need this configuration for running the example. Keep it simple. Just describe the steps to set up and run the example.

File: _posts/2021-05-30-spring-elasticache-redis.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+In this article, we will look at how we can use it to connect our application to [AWS ElastiCache for Redis](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/SelectEngine.html).
+ 
+{% include github-project.html url="https://github.com/thombergs/code-examples/tree/master/aws/spring-cloud-caching-redis" %}
+
+## Why Caching?
+
+Caching is a common technique of temporarily storing a copy of data or result of a computation in memory for quick and frequent access. We use caching primarily for :
+
+1. Improving the throughput of the application.
+2. Prevent overwhelming the application or downstream applications with redundant requests.
```
Comment (thombergs) [inline _posts/2021-05-30-spring-elasticache-redis.md]: This is actually an upstream service, see https://reflectoring.io/upstream-downstream/#upstream-and-downstream-micro-services. This is why I don't like using the terms upstream and downstream: they are confusing.

I changed the wording here.

**Review by thombergs — APPROVED** [review]
@pratikdas @murtuza-ranapur great job! I just changed a few very minor nitpicks.

**Pull request merged**